File: src/Roster.java
import java.util.ArrayList;
import java.util.Random;

/*
Keeps track of the classes we have to pick from. Really, this is an external sort of class for interfacing with some kind of hypothetical database
 */
public class Roster
{
	ArrayList<Integer>[] C;//set of classes
	Random rn;//universal random number generator; use this to guarantee the same roster every time given the same seed
	int numClasses, numStudents;
	
	public Roster(ArrayList<Integer>[] C, int numStudents, Random rn)
	{
		this.C = C;
		this.rn = rn;
		numClasses = C.length;
		this.numStudents = numStudents;
	}
	
	public Roster(int numClasses, int numStudents, Random rn)
	{
		C = new ArrayList[numClasses];
		this.rn = rn;
		this.numClasses = numClasses;
		this.numStudents = numStudents;
		randClasses();
	}
	
	int[] shuffle(int[] a)
	{
		for(int i = 0; i < a.length; i++)
		{
			int ri = rn.nextInt(a.length);
			int temp = a[i];
			a[i] = a[ri];
			a[ri] = temp;
		}
		return a;
	}

	/*
	Generate numClasses random classes to populate the roster. Each class has a uniform number of students chosen
	from the interval [1,sqrt(numStudents)]. This is to avoid massive classes that are both unrealistic and very hard to solve.
	 */
	void randClasses()
	{
		int[] stlist = new int[numStudents];
		for(int i = 0; i < numStudents; i++)
			stlist[i] = i;
		stlist = shuffle(stlist);
		for(int i = 0; i < C.length; i++)
		{
			int classSize = rn.nextInt((int)Math.sqrt(numStudents)-1)+1;
			ArrayList<Integer> nClass = new ArrayList<Integer>(classSize);
			for (int j = 0; j < classSize; j++)
			{
				nClass.add(stlist[j]);
			}
			stlist = shuffle(stlist);
			C[i] = nClass;
		}
	}

	/*
	Print out the roster
	 */
	void show()
	{
		for(int i = 0; i < C.length; i++)
		{
			System.out.print("{");
			for(int j = 0; j < C[i].size()-1; j++)
			{
				System.out.print(C[i].get(j) + ", ");
			}
			System.out.println(C[i].get(C[i].size()-1) + "}");
		}
	}
}

File: src-python/roster.py
from util import *

"""
This class keeps track of the classes we have to pick from. Ideally, this is an external sort of class for interfacing with some kind of hypothetical database
"""
class Roster:

	"""
	C is a set containing all of the classes (lists of student indexes)
	randomState is necessary to make sure that everything uses the same generator. This way we get predictable results
	"""
	def __init__(self, numStudents, C = None, numClasses=None):
		if (C is None) and (numClasses is None):
			raise InputException("Roster class requires the full set of extant classes the number of classes")

		self.C = []
		self.numClasses = None
		self.numStudents = numStudents

		if C is not None:
			self.C = C
			self.numClasses = len(C)
		else:
			self.numClasses = numClasses
			self.randClasses()


	"""
	Generate numClasses random classes to populate the roster. Each class has a uniform number of students chosen
	from the interval [1,sqrt(numStudents)]. This is to avoid massive classes that are both unrealistic and very hard to solve.
	"""
	def randClasses(self):
		stlist = np.array(list(range(self.numStudents)))
		np.random.shuffle(stlist)
		for i in range(self.numClasses):
			classSize = np.random.choice(range(1,int(np.sqrt(self.numStudents)+1)))
			newClass = stlist[:classSize]
			np.random.shuffle(stlist)
			self.C.append(newClass)

	def show(self):
		for cls in self.C:
			print("{",end='')
			for j in range(len(cls)-1):
				print(str(cls[j]) + ", ",end='')
			print(str(cls[-1]) + "}")